Instructor view results: do not auto-load 3 questions #6568

boxin-yang opened the pull request:
<!--
Please follow the naming conventions in the "guidelines for contributing" link above or the pull request may be rejected.

Also, do consider outlining the solution taken as doing so will likely help in the reviewing process
(e.g., when the pull request involves non-trivial changes)
-->

Fixes #6568 

**Outline of Solution**

The auto loading is triggered by $('.ajax_auto').click() . Removing the line will not auto load the results.

boxin-yang commented:
The test is yet updated. Will update later

junhaoyap commented:
@boxin-yang ping us when the tests are updated, that would help us better review and know whether or not the solution worked

sexy-monkey commented:
@junhaoyap  Sure! The problem is in the test that I did not update accordingly. Will tag you again once I finish

boxin-yang commented:
@junhaoyap , I have suppressed two tests for the time being as some weird bugs come out and I think I would need more PR to fix it. Also in this PR, when I tried to play around there are still bugs.(In instructorFeedbackResultsPage and when its sorted by GRP, the default value for panel is still collapse when it is supposed to be expand without auto loading. When I changed the default value to expand, some other unexpected bugs appear with clicking panels)

 Thus I am thinking of putting this PR into a series of more PR into an independent branch. I would need to have merged PR to count for my weekly contribution. What do you think? If you are in Soc sometimes we can discuss about the change :)

damithc commented:
> Thus I am thinking of putting this PR into a series of more PR into an independent branch. I would need to have merged PR to count for my weekly contribution. What do you think? If you are in Soc sometimes we can discuss about the change :)

Normally we expect tests to be in the same PR unless the change is needed urgently (this one isn't).

boxin-yang commented:
@junhaoyap  @wkurniawan07  The code is ready with all the tests changed. Please take a look.

boxin-yang commented:
There are following changes:
1. delete click on ajax_auto and ajax_response_auto at the end of $(document).ready(function() . Thus when page is loaded, the ajax will not start loading.

2. trigger ajax loading by either 
  -- click on collapse/expand panels button that loads both ajax_auto and ajax_response_auto
  -- click on panel header to trigger either ajax_auto/ ajax_response_auto

3. change the initial text of collapse/expand panels button to expand as the page is not loaded by default. change the initial panel button to be invisible as it is not loaded by default.

4. Change the test by adding expandPanels() to pages that were auto-loaded before.

5. Run godMode to change html accordingly.

junhaoyap commented:
Gonna look at this tomorrow, cinderella effect

junhaoyap commented on src/main/webapp/WEB-INF/tags/instructor/results/sectionPanel.tag:
```diff
@@ -26,7 +26,7 @@
                     <div class="col-sm-3">
                         <div class="pull-right">
-                            <a class="btn btn-success btn-xs" id="collapse-panels-button-section-${sectionIndex}" data-toggle="tooltip" title="Collapse or expand all ${isGroupedByTeam? 'team' : 'student'} panels. You can also click on the panel heading to toggle each one individually.">
+                            <a class="btn btn-success btn-xs" id="collapse-panels-button-section-${sectionIndex}" data-toggle="tooltip" title="Collapse or expand all ${isGroupedByTeam? 'team' : 'student'} panels. You can also click on the panel heading to toggle each one individually." style="display: none;">
```
Use a class instead, I am quite sure we have a class which sets the CSS of ```display: none;```, the side effect of using a style like that is that when it is removed for any reason, we get the empty style which is not desirable

junhaoyap commented on src/test/resources/pages/instructorFeedbackResultsAddComment.html:
```diff
@@ -804,7 +804,7 @@
         <div class="col-sm-3">
           <div class="pull-right">
-            <a class="btn btn-success btn-xs" data-original-title="Collapse or expand all team panels. You can also click on the panel heading to toggle each one individually." data-toggle="tooltip" id="collapse-panels-button-section-1" title="">
+            <a class="btn btn-success btn-xs" data-original-title="Collapse or expand all team panels. You can also click on the panel heading to toggle each one individually." data-toggle="tooltip" id="collapse-panels-button-section-1" style="" title="">
```
Like here

junhaoyap commented on src/main/webapp/WEB-INF/tags/instructor/results/sectionPanel.tag:
```diff
@@ -44,7 +44,7 @@
                     <div class="col-sm-3">
                         <div class="pull-right">
-                            <a class="btn btn-success btn-xs" id="collapse-panels-button-section-${sectionIndex}" data-toggle="tooltip" title="Collapse or expand all ${isGroupedByTeam? 'team' : 'student'} panels. You can also click on the panel heading to toggle each one individually.">
+                            <a class="btn btn-success btn-xs" id="collapse-panels-button-section-${sectionIndex}" data-toggle="tooltip" title="Collapse or expand all ${isGroupedByTeam? 'team' : 'student'} panels. You can also click on the panel heading to toggle each one individually."  style="display: none;">
```
Same comment

junhaoyap commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -209,14 +209,21 @@ function toggleCollapse(e, pans) {
-            if ($(panels[idx]).attr('class').indexOf('in') === -1) {
-                setTimeout(showSingleCollapse, 50 * i, panels[idx]);
-                i++;
+        $('.ajax-response-auto').click();
```
This used to be only on a certain page, now that we moved it here, won't there be pages that don't have the element?

junhaoyap replied:
It will not cause any problems if some pages do not have anything selected by ```.ajax-response-auto``` as "clicking on nothing" does nothing

It does create long term maintainance issues when people come around and start asking what it is for, like now :D

boxin-yang replied:
Haha can I create another issue after my PR is merged. I will detailly explain what is going on here and why this is not desirable. I am not sure about how this could be improved. Should I just leave some comments in the code?

junhaoyap replied:
No need to leave comments in the code, just open an issue and reference it to this PR, it seems that different specific pages click on different buttons

damithc replied:
I'd rather not merge this until this is fixed properly. The current solution seems is too hackish. We shouldn't hack-first-fix-later unless it is an urgent issue, which this is not. :-p

junhaoyap commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -209,14 +209,21 @@ function toggleCollapse(e, pans) {
-                setTimeout(showSingleCollapse, 50 * i, panels[idx]);
-                i++;
+        $('.ajax-response-auto').click();
+        var $autoLoadCount = $('.ajax_auto').click();
```
The naming is confusing, the result of ```.click()``` is a jQuery object, not a count, i.e. ```$autoLoad``` or ```$ajaxAutoButton```

junhaoyap commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -209,14 +209,21 @@ function toggleCollapse(e, pans) {
-                i++;
+        $('.ajax-response-auto').click();
+        var $autoLoadCount = $('.ajax_auto').click();
+        var $isAutoLoadingTriggered = $autoLoadCount.length !== 0;
```
And the result of this is a boolean, not a jQuery object, so there is no need to prepend it with ```$```

junhaoyap commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -209,14 +209,21 @@ function toggleCollapse(e, pans) {
+            var i = 0;
+            for (var idx = 0; idx < panels.length; idx++) {
+                if ($(panels[idx]).attr('class').indexOf('in') === -1) {
+                    setTimeout(showSingleCollapse, 50 * i, panels[idx]);
```
I am suspicious of this method ```showSingleCollapse``` as it permanently adds a class of ```in``` and a style of ```height: auto;``` that might cause panels to behave oddly, can you check the function and if it does do all those stuff, please open a new issue and describe that this function should not be doing that

junhaoyap replied:
This a little more advanced, @li-kai @thenaesh can either of you a d.contributor issue regarding this? The original person who wrote this did not understand hoisting well enough, in the issue can we improve this by binding idx to setTimeout instead of maintaining a separate i variable?

http://ignaciothayer.com/post/a-dangerous-example-of-javascript-hoisting/

li-kai replied:
Done! #6712 is opened.

junhaoyap replied:
Yes, leave it unchanged for now, @li-kai has opened an issue regarding it and we have collectively added some good readings for whoever wants to take up refactoring it in the future

junhaoyap replied:
Thanks @li-kai

boxin-yang replied:
Ok will do

junhaoyap commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -209,14 +209,21 @@ function toggleCollapse(e, pans) {
+            isExpandingAll = true;
+            var i = 0;
+            for (var idx = 0; idx < panels.length; idx++) {
+                if ($(panels[idx]).attr('class').indexOf('in') === -1) {
```
Please open a d.firstTimer issue that asks a new contributor to change this to using ```hasClass```, it is way more semantical

junhaoyap commented on src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java:
```diff
@@ -220,7 +220,7 @@ private void testCreateInstructorAction() throws Exception {
         instructorHomePage = AppPage.getNewPageInstance(browser, InstructorHomePage.class);
         instructorHomePage.clickFeedbackSessionViewResultsLink(demoCourseId, "Second team feedback session")
-                          .waitForPageToLoad();
+                          .expandPanels();
```
cc @wkurniawan07 what do you think about these tests? Now that we no longer auto load, I think we can just skip expanding of panels to save on operations and test time, the side-effect is that we will have a lot of HTML changes in this PR and I don't mind that, I would like your opinion

junhaoyap replied:
@wkurniawan07 beep beep

wkurniawan07 replied:
I'm for skipping the expanding parts as well, but it will cause a huge amount of HTML change. We can do that as an extension after this issue.

junhaoyap replied:
Ok, @boxin-yang, open another issue to skip the expanding for the other tests and make sure at least 1 test is testing for either expand / no expand (basically just cover both cases)

boxin-yang commented:
Ready for review @junhaoyap @wkurniawan07

junhaoyap commented:
I have pinged @wkurniawan07 again so let's see when he responds, in the mean time, I understand that adding ```collapse``` class introduces some new HTML styling into the DOM and I am okay with that as it seems like that is just how Twitter's Bootstrap deals with things internally, as long as we do not need to do anything special on top of the framework, I am cool with it

boxin-yang commented:
@junhaoyap The expand functionality is already tested in InstructorFeedbackResultsPageUiTest::testPanelsCollapseExpand(). 
For all the test now I have expanded the panels to minimize the change in html. Actually the majority of change is introduced by adding collapse class. I think the change of HTML is already minimized.

junhaoyap commented:
@boxin-yang the tests are failing 🤔

junhaoyap commented:
@wkurniawan07 understand that you're looking at the accountsDB test that is still failing on this branch after merging with master, otherwise it is good to merge with a few issues going to be oepn

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -211,13 +211,20 @@ function toggleCollapse(e, pans) {
+        var $autoLoad = $('.ajax_auto').click();
+        var isAutoLoadingTriggered = $autoLoad.length !== 0;
+
+        if (!isAutoLoadingTriggered) {
```
What's this whole block doing? Either add a comment to explain or extract into a function with an explanatory name.

boxin-yang replied:
Ok will do

damithc commented on src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java:
```diff
@@ -138,6 +138,26 @@ public void clickShowStats() {
+            clickCollapseExpand();
+            waitForPageToLoad();
+        } catch (NoSuchElementException e) {
+            // For pages have no button, click each of the panel
```
`For pages have no button` needs rephrasing

boxin-yang replied:
noted

damithc commented on src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java:
```diff
@@ -138,6 +138,26 @@ public void clickShowStats() {
+
+    public void expandPanels() {
+        try {
+            // In case the panels are already expanded
```
If the comment is not a full sentence, we usually start with a lower case letter

damithc replied:
Based on this PR, panels should not be already expanded, am I right? In that case the test should fail if the panels are already expanded?

boxin-yang replied:
The function expandPanels() here is trying to expand panels for testing to aim for minimal changes in html in this PR. Some actions cause panels to reload.(For example, when resultPage.displayByGiverRecipientQuestion() is clicked, the page would sort the result panels.)  After reloading, the panels would auto expand in the old implemention.

This PR does not auto expand the panels, thus a lot of clicking expand button is required to be added to the test. To not add too many clicking expand buttons, I include this expandPanels() into every other action that might cause a reload of panels. Checking for whether the panels are already expanded is for cases where<br>
&emsp;&emsp;displayByGiverRecipientQuestion() might be called twice in a row and panels are already expanded before second displayByGiverRecipientQuestion() call. When expandPanels() is called second time in second displayByGiverRecipientQuestion() call, the panels are actually collapsed and the test will fail.

There could be two solutions if we want to change it here:
1. Add in a <b>needToExpandPanels</b> param to each action function in InstructorFeedbackResultsPage (e.g. in displayByGiverRecipientQuestion) to indicate whether we want to expand panels following this action.
2. Delete expandPanels function and call clickCollapseExpand() in the test to expand panels for all cases where panels are reloaded and need to be expanded.

boxin-yang replied:
noted!

damithc replied:
If we keep the current solution (which is just a workaround), are you going to do a proper fix in a separate PR?

boxin-yang replied:
I will have a follow-up PR to change the testing to not expand the panels for most of the testing to save test running time.(Not included in this PR to reduce PR size as not expanding actually changes too many html files)
In that PR I will not include expandPanels into actions that would reload panels.

damithc replied:
OK then. Let me know when this is ready for review again.

boxin-yang commented:
@damithc Ready for review. 

Major changes:

After I look around the js files, I realized that ajax_response_auto is not properly implemented that it does not remove the class ajax_response_auto class after the ajax object is loaded. (As compared to ajax_auto, which remove the class ajax_auto from the element after the ajax loading finishes) . Thus I added the deletion of ajax_response_auto in after the ajax loading is finished for ajax_response_auto.

As for the checking of whether ajax loading should be triggered when the button is clicked, I check first whether the toggleCollapse method is called by the expand panels button. Then I will check whether there is ajax_auto or ajax_response_auto class. (If no such classes, all panels are loaded and no more ajax call is needed)

Also updated some html with godmode for the change that remove ajax_response_auto class.

damithc commented:
I'll take a look and add comments. In the meantime can you get it peer reviewed by someone? Reasons: 1. the code has changed significantly since the previous review 2. it didn't have a peer review to begin with (all PRs should have a peer review)

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -211,11 +211,43 @@ function toggleCollapse(e, pans) {
     var panels = pans || $('div.panel-collapse');
 
     if ($(e).html().trim().startsWith(expand)) {
```
This `if` statement is very long :-p

boxin-yang commented:
@ablyx  Want to review my PR? Msg me if u need any clarification :)

ablyx commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -211,11 +211,43 @@ function toggleCollapse(e, pans) {
+        // panels) and collapse panels button section (button for one panel).
+        var isCollapsePanelsButtonClicked = $(e).is($('#collapse-panels-button'));
+
+        // If expand panels button is clicked and some of panel data is not yet loaded, it will be
```
The comments here are quite long. From what I understand, would this be better:

If expand panel button is clicked and panels have classes ajax_auto or auto-response-auto, it means that the panels do not have data loaded yet. This then loads the data and removes the classes.

boxin-yang replied:
lol that sounds just short and sweet!! Will adapt this one.

ablyx commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -211,11 +211,43 @@ function toggleCollapse(e, pans) {
+            hasAjaxResponseAutoLoading = $ajaxResponseAuto.length !== 0;
+
+            // Ajax_auto is for panels to display normal feedback sessions while ajax-response-auto
+            // is for panels to display students who have no response.
```
Can split the comments for each if block?

boxin-yang replied:
noted

ablyx commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -211,11 +211,43 @@ function toggleCollapse(e, pans) {
-            if ($(panels[idx]).attr('class').indexOf('in') === -1) {
-                setTimeout(showSingleCollapse, 50 * i, panels[idx]);
-                i++;
+        if (!hasAjaxAutoLoading) {
```
Hmm.. can you write some comments to describe why this if check is used?

boxin-yang replied:
Will do. Actually have revised my way of expanding panels

boxin-yang commented:
@damithc  @junhaoyap  Ready for Review

As I have discussed with @wkurniawan07 , I have changed my implementation:

1. Instead of calling all ajax_auto and ajax_response to load panels data when they are not loaded, I check in each of the panel whether ajax call is needed and load accordingly. 

2. If ajax call is needed, the panel is clicked and we no longer need to expand it. This checking is important as the original implementation will expand the panel if it is collapsed.(original implementation does not need to load ajax as it is autoloaded)  However, with the ajax call, the checking for whether the panel is expanded becomes difficult because ajax might click on the panel at a later time when ajax loading is finished.

Please kindly read my changes and leave comments especially on my javascript comments if they are hard to understand or long winded

ablyx commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -209,15 +209,43 @@ function toggleCollapse(e, pans) {
+
+        // Expand each panel if not yet expanded.
+        // When a panel has class ajax_auto or auto-response-auto, the panel data has not been loaded yet.
+        // We will load the data  by ajax and ajax_auto or auto-response-auto class will be removed.
```
There is an extra space here.

ablyx replied:
And can you add a comma after ajax, so that it's 
`We will load the data by ajax, and ajax_auto or ajax-response-auto class will be removed.` Makes what is being loaded and removed clearer.

ablyx replied:
Should be ajax-response-auto. More typos below.

boxin-yang replied:
noted! Good suggestion :)

boxin-yang replied:
lol this error is stupid haha

boxin-yang commented:
@ablyx Hi I have updated some comments and variable name. Hope the code now is easier to understand.

boxin-yang commented:
@ablyx Hi any more updates on this?

ablyx commented:
@boxin-yang sorry, LGTM! Idk if the `mightNeedAjaxLoading` fits the naming convention though

boxin-yang commented:
@damithc  Hi prof, I have made some changes as explained in above conversation. Now the PR is ready for review and I have gotten a peer review from @ablyx . Would you please look up my updates and give some feedbacks .

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -209,15 +209,43 @@ function toggleCollapse(e, pans) {
+        // and collapse panels button section (button for one panel). Class ajax_auto or ajax-response-auto
+        // is found in {@code panels} only if {@link #toggleCollapse} is invoked by clicking on collapse
+        // panels button.
+        var mightNeedAjaxLoading = $(e).is($('#collapse-panels-button'));
```
Why is it called `might`? Does that mean the panel may or may not need ajax loading?

boxin-yang replied:
yup! The panel may or may not need ajax loading. If the panel was clicked before by the instructor, the ajax will load that panel and data for that panel is loaded. If the panel was not clicked before, then we need to load it by ajax first. Thus, for each of the panel, it may or may not need Ajax loading.(The checking here is  to show that if {@code #toggleCollapse} is not called by clicking on collapse panels button, then panel must not need Ajax so we do not need to check for ajax loading for that case )

damithc replied:
In this case (and in the case of `shouldExpandPanels` above) you have a situation like this:
* The code says: `var mustShoot = wifeDrinkingTea;` 
* The explanation: because wife is allergic to tea and if she is drinking tea she must not be the real wife but an alien disguised as wife so therefore I must shoot her before she kills me first. 

I'm exaggerating, but you see my problem right? The variable assignment statement does not tell the whole story and I'm not sure the reader is able to figure out the missing links.  You might need more closer-to-truth variable names and/or explanatory comments. The explanation you gave above (for both cases) gives me a rough idea but I'm not entirely clear either. It's hard to explain this sort of things -- but must keep trying until we get it right.

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -209,15 +209,43 @@ function toggleCollapse(e, pans) {
     var expand = 'Expand';
     var collapse = 'Collapse';
     var panels = pans || $('div.panel-collapse');
+    var shouldExpandPanels = $(e).html().trim().startsWith(expand);
```
What does `$(e).html().trim().startsWith(expand)` actually mean? panel currently collapsed? In that case the variable should be named something like `arePanelsCollapsed`?

boxin-yang replied:
Yup. Here it is checking the text whether to expand the panels. However, the text here does not indicate the status or panels(though by inference we know the panels are supposed to be collapsed). Also, for expand panels button(the one that expands all panels), the button can be clicked regardless of the status of the panels.(Panels can all be collapsed and clicking on the button has no effect other than chaning the status of the button)  Therefore, I think shouldExpandPanels is a better reflection of the purpuse of the $(e).html().trim().startsWith(expand);

boxin-yang commented:
@damithc  Hi Prof I see your analogy. I think I should have explained more for that variable assignment. Now I changed the name to reflect the content assigned and add comments to explain how these two variables are used in the conditional. Hope this time round it would be clearer. Please check my changes :)

boxin-yang commented:
@damithc  Hi Prof, any more suggestion?

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -209,15 +209,44 @@ function toggleCollapse(e, pans) {
                 i++;
             }
         }
+
```
You didn't write this function, so it is not your fault if it is not very easy to understand in the first place. However, you are adding a huge branch to it, so the understandability is getting even worse now. Let's start from the top: What's the high-level algorithm/logic of the `toggleCollapse(e, pans)` function? My naiive expectations is that it is something like this:
```
if already expanded, 
    collapse
else 
    expand
``` 
But I can't seem to find such a high-level logic in the implementation.

boxin-yang replied:
The logic actually has nothing to do with the panels state. It is simply related to the text shown on the button. If the panel button text is "Expand sections", then clicking the button would expand all the panels, regardless of the current state of the panel. (For panels already expanded, do not do anything). Therefore I find it really hard to actually make more changes. My original variable name : 
```shouldExpandPanels = $(e).html().trim().startsWith(expand)```
is named such because that is the logic behind this function: If the button says expand, then expand all panels; else the button would say collapse and toggleCollapse would collapse all the panels.

damithc replied:
So, `isElementAnExpandButton`?

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -209,15 +209,44 @@ function toggleCollapse(e, pans) {
     var panels = pans || $('div.panel-collapse');
+    var isElementAnExpandButton = $(e).html().trim().startsWith(expand);
+
+    if (isElementAnExpandButton) {
```
Now I understand the code up to this line.  Perhaps the next line should be 
`expand(panels)`? That eliminates the need for the comment in line 215.

boxin-yang commented:
@damithc  Hi Prof, I have refactored out both expand and collapse panels and changed the comments accordingly. Please take a look

boxin-yang commented:
@damithc more refactorings are done. Hope now the code is easier to read. Please take a look when you have time

ablyx commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -268,6 +232,54 @@ function toggleCollapse(e, pans) {
+            }
+            if (hasAjaxAutoLoading || hasAjaxResponseAutoLoading) {
+                // When the element needs Ajax loading, it must be collapsed because all expanded panels are
+                // already loaded.As Ajax is triggered by clicking, clicking will expand the collapsed panel.
```
space after full-stop

boxin-yang replied:
noted

ablyx commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -268,6 +232,54 @@ function toggleCollapse(e, pans) {
+            var hasAjaxAutoLoading = $ajaxAuto.length !== 0;
+            var hasAjaxResponseAutoLoading = $ajaxResponseAuto.length !== 0;
+
+            // Ajax loading is triggered by clicking on the element. When clicked, the panel data is loaded
```
`Ajax loading is triggered by clicking on the element, which loads the panel data and removes ajax_auto or ajax-response-auto class from the element` 

I dont think it's necessary to say that it's loaded by Ajax

boxin-yang replied:
true! will change

ablyx commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -268,6 +232,54 @@ function toggleCollapse(e, pans) {
+                $ajaxResponseAuto.click();
+            }
+            if (hasAjaxAutoLoading || hasAjaxResponseAutoLoading) {
+                // When the element needs Ajax loading, it must be collapsed because all expanded panels are
```
I think you just need to mention that clicking expands the panels, that's why the next step is unnecessary.

boxin-yang replied:
Actually I started off just saying clicking expands the panels. But I realize that is not rigorous as clicking **toggles** the panels not expands the panels.So without explaining why the panels are supposed to be collapsed when this ajax clicking happens, some people might wonder what if the panel is expanded when the clicking happens.

boxin-yang replied:
I think we tend to write clicking expands panels because we already know that panels are supposed to be collapsed when we click for ajax loading, and this assumption may not be as obvious to other readers.

ablyx replied:
hmm.. would it be better to say that `Prior to Ajax load, element is collapsed because data is not loaded yet. Clicking toggles both loading of data and expansion of element.`?

Also should it be `element` or `panel`?

boxin-yang commented:
@ablyx thx for the review and I updated some codes. However, I left comments why the current comments might be easier for other people to read, especially when they have no idea about what's going on with Ajax loading. Please leaves comments on how you think about it :) thx

ablyx commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -280,6 +254,33 @@ function collapse(panels) {
+        var mightNeedAjaxLoading = isElementCollapsePanelsButton;
+        expand(panels, mightNeedAjaxLoading);
+        var htmlString = $(e).html();
+        htmlString = htmlString.replace(EXPAND, COLLAPSE);
```
isit possible to refactor the replacement of expand and collapse since it's done twice? might also make things clearer.

boxin-yang replied:
wow didnt think of that and I think it is a good suggestion. Just made the change and I used replaceHTMLString as the method name. Any comment on that?

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,37 +205,79 @@ function updateStatsCheckBox() {
-                setTimeout(showSingleCollapse, 50 * i, panels[idx]);
-                i++;
+/**
+ * Expand each panel. If the panel data is not loaded, load it by Ajax first before expansion.
```
In Java, we normally place the higher level method above the lower level methods. i.e. if foo is calling bar, then we place them as 
```
foo 
bar
```
Can we do the same here?

boxin-yang replied:
Oh I thought js needs to declare the method before calling it. But when I tried to change the position of the code, it is still running well. So I will change all accordingly

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,37 +205,79 @@ function updateStatsCheckBox() {
+    }
+}
+
+function updateHTMLContent(element, from, to) {
```
`update` is a bit too general. `replaceInHtml`? Also, should be `Html` not `HTML`

boxin-yang replied:
hahah Actually I discussed with @ablyx  and he preferred updateHTMLContent. I used update bacause it might be easier to understand that we are doing an update.

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,37 +205,79 @@ function updateStatsCheckBox() {
+    $(element).attr('title', tooltipString).tooltip('fixTitle').tooltip('show');
+}
+
+function toggleCollapse(e, pans) {
```
`toggleExpandCollapse`? It's not fair if you mention only one of the toggle states :-p

damithc replied:
`pans`? Is that a well-known abbreviation? Otherwise give a better name. Same for `e`.

boxin-yang replied:
toggle Collapse. Here the collapse is not referring to the state of the panel, but rather the bootstrap class used. (collapse is a class of the element that we want to toggle)

As for pans and e, I am not very sure with the convention but panels and element might be better

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,37 +205,79 @@ function updateStatsCheckBox() {
+        var isElementCollapsePanelsButton = $(e).is($('#collapse-panels-button'));
+        // We might need to load {@code panels} data by Ajax if {@link #toggleCollapse} is invoked by
+        // clicking on collapse panels button.
+        var mightNeedAjaxLoading = isElementCollapsePanelsButton;
```
If ajax loading is something to do with expanding, it can be dealt with inside the `expand` method.

boxin-yang replied:
yup I also thought about this and updated my local repo.

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,37 +205,79 @@ function updateStatsCheckBox() {
+function expand(panels, needToCheckForAjaxLoading) {
+    var i = 0;
+    for (var idx = 0; idx < panels.length; idx++) {
+        if (needToCheckForAjaxLoading) {
```
Instead of this
```
function foo(){
  if(x){
    //very
    //long
    //block
  }
}
```
can do this (i.e. try to avoid nesting the 'happy path' of the code) ?
```
function foo(){
  if(!x){
    return;
  }
  //very
  //long
  //block
}

boxin-yang replied:
Unfortunately we cannot do it here. The else block must be executed after the very long block because the very long block might **continue** the for loop.  the code following the long block is dependent on the long block.

boxin-yang commented:
@damithc  I have made changes accordingly. As for the new variable name panelsDefined, it is a walkaround with eslint checking. I cannot assign
```var panels= panels || $('div.panel-collapse'); // reassignment```
```panels= panels || $('div.panel-collapse'); //should not vary param```
Avoiding conllision with local variable names might be the reason to use pans instead of panels in the original code.

Also, just FYI: 
```panels= panels || $('div.panel-collapse');```
means if the panels are undefined, panels = $('div.panel-collapse')

boxin-yang commented:
@damithc  ready for review

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,82 @@ function updateStatsCheckBox() {
-        $(e).html(htmlString);
-        var tooltipString = $(e).attr('data-original-title').replace(expand, collapse);
-        $(e).attr('title', tooltipString).tooltip('fixTitle').tooltip('show');
+function toggleCollapse(element, panels) {
```
Add header comments to the new functions

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,82 @@ function updateStatsCheckBox() {
-        $(e).attr('title', tooltipString).tooltip('fixTitle').tooltip('show');
+function toggleCollapse(element, panels) {
+    var EXPAND = 'Expand';
+    var COLLAPSE = 'Collapse';
```
The constant names should be more explanatory

boxin-yang replied:
noted

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,82 @@ function updateStatsCheckBox() {
+    }
+}
+/**
+ * Expand each panel. If the panel data is not loaded, load it by Ajax first before expansion.
```
`Exapands`

damithc replied:
Need to describe the `element` parameter, as it is not obvious.

boxin-yang replied:
noted

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,82 @@ function updateStatsCheckBox() {
+ * Expand each panel. If the panel data is not loaded, load it by Ajax first before expansion.
+ */
+function expand(panels, element) {
+    var isElementCollapsePanelsButton = $(element).is($('#collapse-panels-button'));
```
Previously this element was recognized as an 'expand button'; what's this about?

boxin-yang replied:
The previous variable was just to say the button text appeared in html starts with Expand. Here it means the element is of id collapse-panels-button

damithc replied:
Why do we care that it's collapse-panels-button? i.e. what's the implication?

boxin-yang replied:
```var isElementCollapsePanelsButton = $(element).is($('#collapse-panels-button'));
    // Ajax class is only found in {@code panels} if the element is collapse panels button.
    var mightNeedAjaxLoading = isElementCollapsePanelsButton;
```
If it is a collapse-panels-button, we will need to check for ajax, else we do not need to check.

damithc replied:
How about this?
```js
// Expand/collapse buttons on AJAX-loaded panels have a collapse-panels-button class
var isAjaxLoadedPanel = $(element).is($('#collapse-panels-button'));
```

Is it correct to write like that?

damithc replied:
Feel free to elaborate/improve.

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,82 @@ function updateStatsCheckBox() {
+            // Expand the panel if the panel is collapsed at the moment
+            setTimeout(showSingleCollapse, 50 * i, panels[idx]);
+            i++;
+        }
```
Is it possible to extract a method that expands only one panel and call that method in a loop?

boxin-yang replied:
That method would not improve the readability as everything inside the while loop is to expand one panel. Even if we extract the method, the method still has to contain all the Ajax checking and expand panels checking.

damithc replied:
Extract it anyway. Expanding a panel is a separate unit of work and should be a separate function.

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,82 @@ function updateStatsCheckBox() {
+function collapse(panels) {
+    var i = 0;
+    for (var inx = 0; inx < panels.length; inx++) {
+        if ($(panels[inx]).attr('class').indexOf('in') !== -1) {
```
What's the meaning of this condition? Introduce explanatory variable?

boxin-yang replied:
Same as before, this is intended to be a first timer issue

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,82 @@ function updateStatsCheckBox() {
-        var tooltipStr = $(e).attr('data-original-title').replace(collapse, expand);
-        $(e).attr('title', tooltipStr).tooltip('fixTitle').tooltip('show');
+
+        if ($(panels[idx]).attr('class').indexOf('in') === -1) {
```
What's the meaning of this condition? Introduce explanatory variable?

damithc replied:
Or extract into a small method with an explanatory name.

boxin-yang replied:
This is left as a first timer issue. This is just checking for whether the element has in class. (In bootstrap, the collapse class will becomes collapse class + in class when clicked to hide the panel). I purposely did not change this as this is intended to be a first timer issue.

damithc replied:
As we are already deep into this, let's do all we can to clean up this part of the code. No need to leave out anything. :-)

boxin-yang replied:
ok !

boxin-yang commented:
@damithc  Hi prof, I have made more refactoring with more comments and extracted out some classes. It is ready for review , please take a look.

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,96 @@ function updateStatsCheckBox() {
+/**
+ * Expand or collapse all panels when collapse/expand panels button is clicked
+ * @param {DOM} element - The element that invoked {@code #expandOrCollapsePanels}
+ * @param {DOM} panels - The panels element to be changed.
```
It's a bit odd to hear `panels element` because first word is plural and second one is singular. Is it a `panel element possibly containing more panels inside it`?

boxin-yang replied:
It should be panels elements

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,96 @@ function updateStatsCheckBox() {
+ */
+function expandPanels(panels, element) {
+    // Expand/collapse buttons on AJAX-loaded panels have a collapse-panels-button class
+    var isAjaxLoadedPanel = $(element).is($('#collapse-panels-button'));
```
Add a blank line at 236 to indicate that the comment at line 234 applies to line 235 only. Otherwise it applies to the entire code block.

boxin-yang replied:
Oh I see

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,96 @@ function updateStatsCheckBox() {
+    var i = 0;
+    for (var idx = 0; idx < panels.length; idx++) {
+        expandPanel(panels[idx], 50 * i, isAjaxLoadedPanel);
+        i++;
```
Do you need `i`? Can use `idx` instead?

boxin-yang replied:
yup. This was the legacy of refactoring

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,96 @@ function updateStatsCheckBox() {
+            continue;
+        }
+        // collapse this panel
+        setTimeout(hideSingleCollapse, 100 * i, panels[idx]);
```
Define magic number 100 as a constant?

boxin-yang replied:
ok

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,96 @@ function updateStatsCheckBox() {
+            // When the element has Ajax class, its panel is in collapsed state. Clicking will expand
+            // its panels.
+            return;
         }
```
Extract this whole block as a function. It has a distinct job, which to populate the panel using an Ajax call.

boxin-yang replied:
ok

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,96 @@ function updateStatsCheckBox() {
+        }
+        if (hasAjaxResponseAutoLoading) {
+            $ajaxResponseAuto.click();
+        }
```
Can both of these happen or only one? In the latter case it should be an `else if`.

boxin-yang replied:
In the correct implementation only one can happen

damithc replied:
In that case, this function can be simplified a bit more.

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,96 @@ function updateStatsCheckBox() {
+        var $ajaxAuto = $(panel).parent().children('.ajax_auto');
+        var $ajaxResponseAuto = $(panel).parent().children('.ajax-response-auto');
+        var hasAjaxAutoLoading = $ajaxAuto.length !== 0;
+        var hasAjaxResponseAutoLoading = $ajaxResponseAuto.length !== 0;
```
What's the difference between these two?

boxin-yang replied:
The original code base implemented ajax_auto to retrieve for normal panels data while ajax_response_auto retrieves data for people who have not responded the questions. Do you think I should explain it here. (I deleted the explanation as I thought it is not relevant to the logic here but more of a good to know information)

damithc replied:
Name them to reflect these two cases correctly, instead of repeating the implementation detail as the variable name. After that, add comments to explain further. Basically, the reader should not have to ask the questions I'm asking.

Anyhow, may be use one variable? Replace `xxx` with a better name.

```js
var $xxx = $(panel).parent().children('.ajax_auto');
if ($xxx.length === 0) {
    $xxx = $(panel).parent().children('.ajax-response-auto');
}
```

damithc replied:
My syntax above may be wrong; I'm not a JS guy.

boxin-yang replied:
The reason I put ajax_auto and ajax-response-auto as two separate variables  and used hasAjaxAutoLoading  and hasAjaxResponseAutoLoading  is for better readability. (For people not familiar with Jquery).

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,96 @@ function updateStatsCheckBox() {
+    }
+}
+
+function replaceInHtml(element, from, to) {
```
header comment missing

boxin-yang replied:
noted

boxin-yang commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,96 @@ function updateStatsCheckBox() {
-        var tooltipString = $(e).attr('data-original-title').replace(expand, collapse);
-        $(e).attr('title', tooltipString).tooltip('fixTitle').tooltip('show');
+/**
+ * Expand or collapse all panels when collapse/expand panels button is clicked
```
noted

boxin-yang commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,96 @@ function updateStatsCheckBox() {
+    var STRING_EXPAND = 'Expand';
+    var STRING_COLLAPSE = 'Collapse';
+    // {@code panels} is undefined when the method is called by clicking on collapse panels button.
+    var panelsDefined = panels || $('div.panel-collapse');
```
Thx for the suggestion

boxin-yang commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,96 @@ function updateStatsCheckBox() {
+ * Expands all panels.
+ * @param {DOM} element - The element that was clicked to invoke {@code #expandOrCollapsePanels}
+ */
+function expandPanels(panels, element) {
```
noted!

boxin-yang commented:
@damithc Refactored as suggested, please take a look

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,110 @@ function updateStatsCheckBox() {
+
+    if (isElementAnExpandButton) {
+        expandPanels(element, targetPanels);
+        replaceInHtml(element, STRING_EXPAND, STRING_COLLAPSE);
```
Shouldn't this line be inside the expandPanels function? Apparently, it's part of expanding panels.

boxin-yang replied:
Sorry, should have used a better name replaceButtonHtmlAndTooltip. This function is to expand buttons not panels

damithc replied:
Isn't the button part of the panel? and hence updating the button should be part of expanding the panel?

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,110 @@ function updateStatsCheckBox() {
+function expandPanels(element, panels) {
+    // Expand/collapse buttons on AJAX-loaded panels have a collapse-panels-button class
+    var areAjaxLoadedPanels = $(element).is($('#collapse-panels-button'));
+    var TWENTIETH_SECOND_TO_MILLI_SECOND = 50;
```
What's the meaning of this constant?

damithc replied:
I can't figure out how this constant got named like this 😛

boxin-yang replied:
thats the meaning of 50 in the code. The timeout value * 50 changes the unit of timeout from 1/20 of a second to millisecond.

damithc replied:
That's not the 'meaning'; that's the 'mechanism' of how it was calculated. You tend to name things based on mechanism when you should name things based on the meaning/intent instead. Is it the `TIME_ALLOWED_FOR_ONE_PANEL_EXPANDING` ?

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,110 @@ function updateStatsCheckBox() {
 }
 
+/**
+ * Replace the Html content after the panel expands/collapses.
```
For the third time, `Replaces`

damithc replied:
What html content exactly? Need to explain more, for example, the fact that the tooltip will be updated.
Is this function for panels only? The name doesn't suggest any such restriction but the header comment does.

boxin-yang replied:
this function is for buttons only. Updated

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,110 @@ function updateStatsCheckBox() {
+}
+
+/**
+ * Expands all panels.
```
Also, loads data if missing, right?

boxin-yang replied:
noted

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,110 @@ function updateStatsCheckBox() {
+    var $ajaxElement = $(panel).parent().children('.ajax_auto');
+    if ($ajaxElement.length === 0) {
+        $ajaxElement = $(panel).parent().children('.ajax-response-auto');
+    }
```
What's an `ajax element`? Is the reader supposed to understand this term intuitively?
Did you mean `elementToClickForAjaxLoading`?

boxin-yang replied:
updated to $elementWithAjaxClass

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,110 @@ function updateStatsCheckBox() {
+        $ajaxElement = $(panel).parent().children('.ajax-response-auto');
+    }
+
+    var hasAjaxElement = $ajaxElement.length !== 0;
```
Can this be false at this point?

boxin-yang replied:
Yes. If no Ajax class is found, the hasAjaxElement will be false. This happens when clicking collapse panels button for more than once.

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,110 @@ function updateStatsCheckBox() {
+        }
+    }
+    // expand this panel
+    setTimeout(showSingleCollapse, timeOut, panel);
```
How about this?
```js
if (isAjaxLoadedPanel) {
    checkAndLoadPanelByAjax(panel);
    // panel is expanded automatically after ajax loading
} else {
    // expand this panel without ajax loading
    setTimeout(showSingleCollapse, timeOut, panel);
}
```

boxin-yang replied:
this is not the intention of the function. The function is hard to refactor for the fact that if checkAndLoadPanelByAjax returns false(which means the panel data is already loaded), we still need to setTimeout(showSingleCollapse, timeOut, panel) to expand the panel.  setTimeout(showSingleCollapse, timeOut, panel) is dependent on the result of checkAndLoadPanelByAjax

damithc replied:
That's the problem with doing multiple things in a function. Names such as `foodAndBar` is always not a good sign.
narrow it down to just one job
`getElementToClickForAjaxLoading()`


```js
var $elementToClickForAjaxLoading = getElementToClickForAjaxLoading(panel);

if ($elementToClickForAjaxLoading !== 0) { 
    $elementToClickForAjaxLoading.click();
    // panel is expanded automatically after ajax loading
} else {
    // expand this panel without ajax loading
    setTimeout(showSingleCollapse, timeOut, panel);
}

boxin-yang commented:
Responded to some of your questions and made updates if applicable.  As for TWENTIETH_SECOND_TO_MILLI_SECOND, the better name I can think of is BASE_TIME_OUT_UNIT_IN_MILLI_SECOND. Hope that would make more sense. 
@damithc  Ready for review

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,109 @@ function updateStatsCheckBox() {
+
+    if (isElementAnExpandButton) {
+        expandPanels(element, targetPanels);
+        replaceButtonHtmlAndTooltipText(element, STRING_EXPAND, STRING_COLLAPSE);
```
Isn't the button part of panel, and hence, updating the button is part of expanding the panel?

boxin-yang replied:
Erm they are different things in different div. Button is more like a remote to all the panels and they are not related at all. (If you open feedback result page,  you would see Expand Questions button and each question panel as not really related
![screen shot 2017-03-22 at 11 59 25 pm](https://cloud.githubusercontent.com/assets/16921967/24207366/9e1985d8-0f5b-11e7-9b0b-6923ffc455d1.png)
)

boxin-yang commented:
@damithc  Ready for review

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,102 @@ function updateStatsCheckBox() {
+}
+
+/**
+ * @return {DOM} the element that has class ajax_auto or ajax-response-auto(not both). The element can be
```
I think `not both` has no relevance here. The code works the same way even if it has both.

boxin-yang replied:
Here is more of a FYI for future reader or maintainer. (So they know that at most one class exists and the assignment of $elementToClickForAjaxLoading will not need to handle the case with both classes present) And I think this is good to know, else people might be confused.

damithc replied:
Header comment is meant for the users of the function. Comments to the maintainers should be given inside the function.

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,102 @@ function updateStatsCheckBox() {
+
+/**
+ * @return {DOM} the element that has class ajax_auto or ajax-response-auto(not both). The element can be
+ *         clicked to load the {@param panel} content by Ajax.
```
Can be phrased as `@return {DOM} the element that needs to be clicked to do an AJAX-loading of data to the panel, identified by the presence of the {@code ajax_auto} or {@code ajax-response-auto} class attached to the element.`

Not sure if the `{@code }` annotation works for JS though. Also not sure the two things are classes or something else.

boxin-yang replied:
{@code string} is used to indicate the string is code. (according to https://google.github.io/styleguide/jsguide.html) .  Two things are classes of the element

boxin-yang replied:
However, I do not think we should use {@code ajax_auto} here as this ajax_auto does not refer to any code.

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,102 @@ function updateStatsCheckBox() {
+/**
+ * Expands or collapses all panels when collapse/expand panels button is clicked.
+ * @param {DOM} element - The element that was clicked to invoke {@code #expandOrCollapsePanels}
+ * @param {DOM} panels - The panels to be expanded/collapsed. not defined if {@param element} is
```
`not` -> `Not`

boxin-yang replied:
noted

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,102 @@ function updateStatsCheckBox() {
+function expandOrCollapsePanels(element, panels) {
+    var STRING_EXPAND = 'Expand';
+    var STRING_COLLAPSE = 'Collapse';
+    var targetPanels = panels || $('div.panel-collapse');
```
Need to explain what's going on here and why it is needed.

boxin-yang replied:
ok

boxin-yang commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,102 @@ function updateStatsCheckBox() {
-        $(e).attr('title', tooltipString).tooltip('fixTitle').tooltip('show');
+/**
+ * Expands or collapses all panels when collapse/expand panels button is clicked.
+ * @param {DOM} element - The element that was clicked to invoke {@code #expandOrCollapsePanels}
```
yup.

boxin-yang commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,102 @@ function updateStatsCheckBox() {
+ */
+function expandPanels(element, panels) {
+    // The expand/collapse button on AJAX-loaded panels is of id collapse-panels-button
+    var areAjaxLoadedPanels = $(element).is($('#collapse-panels-button'));
```
ok

boxin-yang commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,102 @@ function updateStatsCheckBox() {
+    var STRING_EXPAND = 'Expand';
+    var STRING_COLLAPSE = 'Collapse';
+    var targetPanels = panels || $('div.panel-collapse');
+    var isElementAnExpandButton = $(element).html().trim().startsWith(STRING_EXPAND);
```
ok

boxin-yang commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,38 +205,102 @@ function updateStatsCheckBox() {
+function expandPanels(element, panels) {
+    // The expand/collapse button on AJAX-loaded panels is of id collapse-panels-button
+    var areAjaxLoadedPanels = $(element).is($('#collapse-panels-button'));
+    var BASE_TIME_OUT_UNIT_IN_MILLI_SECOND = 50;
```
noted!

boxin-yang commented:
@damithc Ready for review

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,107 @@ function updateStatsCheckBox() {
+    var STRING_COLLAPSE = 'Collapse';
+    // {@param panels} is not defined when @param expandCollapseButton} is collapse panels button. We
+    // need to define corresponding {@code targetPanels}.
+    var targetPanels = panels || $('div.panel-collapse');
```
So this is only applicable to the `else` branch? In that case move it there.

boxin-yang replied:
This is applicable to all branch. This is about which button was clicked, not whether button is in expand or collapse mode. and this is the javascript way of assigning a default value to the variable not initiated.

damithc replied:
I thought there is only one button that is being used to expand and collapse? What are other button?
![image](https://cloud.githubusercontent.com/assets/1673303/24251761/3a9427e6-1016-11e7-8784-99e6b6d505a9.png)

damithc replied:
Alternatively, why can't the collapse button specify the panel just like the expand button? That would simplify this function.

boxin-yang replied:
There are more than one button.  Each panel when view Group by a mode other than (Group by - Question), has a button at the right side of the panel to expand and collapse all panels(all panels in that panel)

boxin-yang replied:
Here it is not collapse button that specifies the panel. It is the button with id collapse-panels-button( a very bad name, but it has nothing to do with whether button is in expand or collapse mode).  The button id does not change with panel state

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,107 @@ function updateStatsCheckBox() {
+
+    var isButtonInExapandMode = $(expandCollapseButton).html().trim().startsWith(STRING_EXPAND);
+    if (isButtonInExapandMode) {
+        // The expand/collapse button on AJAX-loaded panels is of id collapse-panels-button
```
`is of id` -> `has id`?

boxin-yang replied:
ok

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,107 @@ function updateStatsCheckBox() {
+}
+
+/**
+ * Expands {@param panel}. If the panel data is not loaded, load it by Ajax.
```
`{@param panel}` or `{@code panel}`. In java `@param` is used only when documenting a param, not referring to a param.

boxin-yang replied:
ok changed all

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,107 @@ function updateStatsCheckBox() {
+    }
+
+    if (isAjaxLoadedPanel) {
+        // Might need to load the panel data by Ajax.
```
This comment can be at the end of line 254

boxin-yang replied:
ok

boxin-yang commented:
@damithc Ready for review

damithc commented on src/main/webapp/js/instructorFeedbackResults.js:
```diff
@@ -205,40 +205,106 @@ function updateStatsCheckBox() {
+ * Expands all panels. Loads panel data if missing.
+ */
+function expandPanels(panels, areAjaxLoadedPanels) {
+    var BASE_TIMEOUT_UNIT_IN_MILLI_SECONDS = 50;
```
Um, why collapse is give twice as long compared to expand? Any idea?

boxin-yang replied:
One possible reason is that visual effect of expanding is not really shown most of the time, but collapse is shown as the collapsed panels fit in one page. I cannot think of other reasons though.

damithc commented:
@boxin-yang one more [query](https://github.com/TEAMMATES/teammates/pull/6608#discussion_r107678334) awaiting your response.

damithc commented:
In the meantime can @wkurniawan07 have a quick look through the current version?

boxin-yang commented:
@damithc  The query was addressed yesterday. Ready for review

damithc commented:
> The query was addressed yesterday.

It was? I can't see the answer in the thread. Here's the question again:

![image](https://cloud.githubusercontent.com/assets/1673303/24298704/cbcb1d16-10e1-11e7-8b61-c405d20ed072.png)

boxin-yang commented:
@damithc  Sorry didnt say that comment. Just replied and hope it clarifies

boxin-yang commented:
@damithc  Just a screenshot to show you where are the rest of the buttons: 
![screen shot 2017-03-24 at 11 37 40 pm](https://cloud.githubusercontent.com/assets/16921967/24301606/f9b7b028-10ea-11e7-91cc-5e72484e8742.png)
And all the buttons are bond to expandOrCollapsePanels. Only the button at the top might trigger ajax.

wkurniawan07 commented:
I'll help de-conflicting with the latest `master` later.

boxin-yang commented:
@damithc  Ready for review

wkurniawan07 commented on src/main/webapp/WEB-INF/tags/instructor/results/sectionPanel.tag:
```diff
@@ -44,7 +44,7 @@
                     <div class="col-sm-3">
                         <div class="pull-right">
-                            <a class="btn btn-success btn-xs" id="collapse-panels-button-section-${sectionIndex}" data-toggle="tooltip" title="Collapse or expand all ${isGroupedByTeam? 'team' : 'student'} panels. You can also click on the panel heading to toggle each one individually.">
+                            <a class="btn btn-success btn-xs collapse" id="collapse-panels-button-section-${sectionIndex}" data-toggle="tooltip" title="Collapse or expand all ${isGroupedByTeam? 'team' : 'student'} panels. You can also click on the panel heading to toggle each one individually."">
```
There is an extra `"` at the end.
